### PR TITLE
expanded pbkdf2-hmac-sha1 opencl to 179 bytes, and kernel is unlimite…

### PR DESCRIPTION
**Pull request by jfoug**
Tests fine on pbkdf2_hmac_sha1-opencl.  Also added many tests, to know that multi limb code is working.  The kernel should now handle unlimited length salts (controlled by the CPU format setup).

### CODE REVIEW

**Comment by jfoug** [comment]
For now, I created a new define  PBKDF2_32_MAX_SALT_SIZE_SHA1   The older define used by all 32 bit hash types was PBKDF2_32_MAX_SALT_SIZE    It was set to 115 (max 2 limb length).  I changed the SHA1 in opencl to handle any length. But we still have to control the structure itself, within the CPU code.

All of the CPU pbkdf2-* stuff in the header files handles salts of any length (it simply uses CTX methods to hash salts > 64 bytes).  This code is ONLY done at setup for pbkdf2, and then re-used within the internal iterations.

ONCE we get all other 32 bit hash pbkdf2-hmac-* updated like what was done in sha1 (to handle unlimited length salts IN the kernel), then we can scrap the   PBKDF2_32_MAX_SALT_SIZE_SHA1 and simply change  PBKDF2_32_MAX_SALT_SIZE to be 179   (3 limb max salt size).  OR we can keep the SHA1 longer (since it is used IRL) and keep the others at 2 blocks. However, we still SHOULD change the kernels to handle any length salts, so later if we change the CPU part of the formats, the kernels are happy to work properly.

File: src/opencl_pbkdf2_hmac_sha1.h [hunk 1]
```diff
@@ -26,12 +28,28 @@ typedef struct {
...
+//	unsigned char salt[55];
+//} pbkdf2_salt;
+//#elif defined (OPENCL_PBKDF2_HMAC_SHA1_2_LIMB)
+//typedef struct {
+//	unsigned int  length;
+//	unsigned int  outlen;
+//	unsigned int  iterations;
+//	unsigned char salt[115];
+//} pbkdf2_salt;
+//#elif defined (OPENCL_PBKDF2_HMAC_SHA1_3_LIMB)
```
Comment (magnumripper) [inline src/opencl_pbkdf2_hmac_sha1.h]: Perhaps you should drop the commented-out stuff? Or is it better left there?

**Comment by kholia** [comment]
I am going to test this code by implementing `OpenBSD-SoftRAID-opencl` on top of it.

Some quick comments,

1. DOS line endings have made it to `pbkdf2_hmac_common_plug.c` file.

2. Trailing space added in `pbkdf2_hmac_sha1_kernel.cl` file.

You can fix these minor problems, then do `git commit --amend`, followed by `git push -f`

Thanks!

**Comment by kholia** [comment]
> I am going to test this code by implementing OpenBSD-SoftRAID-opencl on top of it.

Done, `opencl_openbsdsoftraid_fmt_plug.c` works just fine without making any changes. Nice work!

File: src/opencl_pbkdf2_hmac_sha1.h (hunk 1 above)
Comment (jfoug) [reply src/opencl_pbkdf2_hmac_sha1.h]: I left it there, as a 'guide'.  I think we might see some performance gains (possibly not), by having the formats define that they only need 51 byte salt, (or some other size).

We can take it out for now, if you like, but I would rather the comments be there, since that part of the code is 'done', then later we can add the defines in the proper places within the formats.

But since this is only the salt size, it may not be that big of an impact on the formats at all. I think the only time there is any data transfer to GPU is when the salt is set.

**Comment by jfoug** [comment]
The code you sent me did not work for openbsdsoftraid.  Possibly it is my GPU (there are some other formats which have failures), or it could be code you sent that was not quite ready.  

BUT the opencl version of pbkdf2-hmac-sha was the test bed I used. Until it was working properly with all the carefully selected salt lengths, I was not going to check anything in.  We might add a few more salt lengths  (123-130) to test boundaries on 2 limb.  I did test them on 1 limb (and they caught a bug I had in code), so testing boundary at 2 limb is not quite as critical, but it probably should be added.

**Comment by jfoug** [comment]
> Some quick comments,

Corrected.

**Comment by jfoug** [comment]
Does anything use pbkdf2_kernel.cl  ?  (looks like mscash2 uses it). Is there any reason why?  Is that better or worse code?  If better, then why not used in all formats?

File: src/opencl_pbkdf2_hmac_sha1.h (hunk 1 above)
Comment (magnumripper) [reply src/opencl_pbkdf2_hmac_sha1.h]: Lets keep it. I'll look into some clean-up and optimizations later, letting GPU side know actual max-size of salt and adopt at compile time. This is not in the hot loop so we'll probably not see much gain, but anyway.

**Comment by kholia** [comment]
> The code you sent me did not work for `OpenBSD-SoftRAID-opencl`. Possibly it is my GPU (there are some other formats which have failures), or it could be code you sent that was not quite ready.

Hmm, `OpenBSD-SoftRAID-opencl ` fails on Travis CI too. I will look into it.

**Comment by magnumripper** [comment]
> Does anything use pbkdf2_kernel.cl ? (looks like mscash2 uses it). Is there any reason why? Is that better or worse code? If better, then why not used in all formats?

That one is Sayantan's. Nothing wrong with it (and it's not any better) except it's more or less unreadable to me and his formats are so different from all others, I don't dare trying to unify them.

Oh, and (I can't remember) perhaps that's the one single format we have that is capable of using several GPU's without forking?

**Comment by kholia** [comment]
> Hmm, OpenBSD-SoftRAID-opencl fails on Travis CI too. I will look into it.

Here is a patch to fix the problem,

``` diff
diff --git a/src/opencl_openbsdsoftraid_fmt_plug.c b/src/opencl_openbsdsoftraid_fmt_plug.c
index 6ca31d1..7680687 100644
--- a/src/opencl_openbsdsoftraid_fmt_plug.c
+++ b/src/opencl_openbsdsoftraid_fmt_plug.c
@@ -294,10 +294,11 @@ static int crypt_all(int *pcount, struct db_salt *salt)
                        unsigned char unmasked_keys[OPENBSD_SOFTRAID_KEYLENGTH * OPENBSD_SOFTRAID_KEYS];
                        unsigned char hashed_mask_key[20];
                        AES_KEY akey;
+                       int k;
 
                        AES_set_decrypt_key((unsigned char*)output[index].dk, 256, &akey);
-                       for (j = 0; j < (OPENBSD_SOFTRAID_KEYLENGTH * OPENBSD_SOFTRAID_KEYS) / 16;  j++) {
-                               AES_decrypt(&cur_salt->masked_keys[16*j], &unmasked_keys[16*j], &akey);
+                       for (k = 0; k < (OPENBSD_SOFTRAID_KEYLENGTH * OPENBSD_SOFTRAID_KEYS) / 16;  k++) {
+                               AES_decrypt(&cur_salt->masked_keys[16*k], &unmasked_keys[16*k], &akey);
                        }
 
                        /* get SHA1 of mask_key */
```

This problem was not caught in my macOS VM as there is no OpenMP support there.
